src/Model/PhasorDynamics/Bus/BusSignal.hpp

```diff
@@ -0,0 +1,127 @@
+
+#pragma once
+
+#include <Model/PhasorDynamics/BusBase.hpp>
+
+namespace GridKit
+{
+  namespace PhasorDynamics
+  {
+
+    // TODO change base class so there is no Vr and Vi
+
+    template <class ScalarT, typename IdxT>
+    class BusSignal : public BusBase<ScalarT, IdxT>
+    {
+      using BusBase<ScalarT, IdxT>::size_;
+      using BusBase<ScalarT, IdxT>::y_;
+      using BusBase<ScalarT, IdxT>::yp_;
+      using BusBase<ScalarT, IdxT>::yB_;
+      using BusBase<ScalarT, IdxT>::ypB_;
+      using BusBase<ScalarT, IdxT>::f_;
+      using BusBase<ScalarT, IdxT>::fB_;
+      using BusBase<ScalarT, IdxT>::tag_;
+
+    public:
+      using real_type = typename BusBase<ScalarT, IdxT>::real_type;
+      using DataT     = BusData<real_type, IdxT>;
+
+      BusSignal(const DataT& data);
+
+      ~BusSignal()
+      {
+      }
+
+      virtual int allocate() override;
+      virtual int tagDifferentiable() override;
+      virtual int initialize() override;
+      virtual int evaluateResidual() override;
+      virtual int evaluateJacobian() override;
+
+      /**
+       * @brief A one-time initialization function
+       * that can be called by any.
+       *
+       * @return state value of signal
+       */
+      void initial_value(ScalarT value) override
+      {
+
+        y_[0] = value;
+        // if(is_initialized_)
+        //{
+        //   std::cout << "ERROR!";
+        // }
+        // else{
+        //   y_[0] = value;
+        //   is_initialized_ = true;
+        // }
+      }
+
+      /**
+       * @brief A read only accessor to the signal state
+       *
+       * @return state value of signal
+       */
+      ScalarT& read() override
+      {
+        return y_[0];
+      }
+
+      /**
+       * @brief A write operation only available
+       * to the source signal
+       *
+       * @param value The value of the signal
+       */
+      void write(ScalarT& value) override
+      {
+        f_[0] += value;
+      }
```

Comment:
I believe this operation should never be allowed. It is _only the solver_ who should write the new values to the system variables. This operation may alter solution values and consequently system residual and Jacobian evaluation.

A better implementation would be for the source component to have a reference to the signal value and to use it in equations directly. A potential issue with signal values being stored in signal buses is that signal connections are decided at runtime. This means that generator, for example, would have to be able to configure its equations at runtime to use values from signal buses or its local variables based on user-chosen connectivity. This may also complicate computing derivatives of component equations at _compile time_.